core/extensions/impl/crypto_extension.cpp

```diff
@@ -168,4 +174,75 @@ namespace kagome::extensions {
 
     memory_->storeBuffer(out_ptr, common::Buffer(hash));
   }
+
+  runtime::PointerSize CryptoExtension::ext_crypto_secp256k1_ecdsa_recover_v1(
+      runtime::WasmPointer sig, runtime::WasmPointer msg) {
+    static auto make_result = [&](auto &&value) {
+      boost::optional<crypto::Secp256k1UncompressedPublicKey> result = value;
+      return common::Buffer(scale::encode(result).value());
+    };
+    static auto error_result = make_result(boost::none);
+
+    crypto::Secp256k1Signature signature{};
+    crypto::Secp256k1Message message{};
+
+    const auto &sig_buffer = memory_->loadN(sig, signature.size());
+    const auto &msg_buffer =
+        memory_->loadN(msg, crypto::Secp256k1Message::size());
+    std::copy_n(sig_buffer.begin(), signature.size(), signature.begin());
+    std::copy_n(
+        msg_buffer.begin(), crypto::Secp256k1Message::size(), message.begin());
+
+    auto &&public_key =
+        secp256k1_provider_->recoverPublickeyUncompressed(signature, message);
+    if (!public_key) {
+      logger_->error("failed to recover uncompressed secp256k1 public key");
+      return memory_->storeBuffer(error_result);
+    }
+    auto &&encoded = scale::encode(public_key.value());
```

Comment:
Actually our encode function never returns error. I don't know why we decided to return result there instead of vector. You can save tons of LOC by just taking value from encode result

Reply:
Let's make task for refactoring. This result type really useless.

Reply:
Created a jira task [PRE-450](https://soramitsu.atlassian.net/browse/PRE-450)